### ISSUE TITLE
[Potential Bug] Use mip level to query density grid.

### ISSUE THREAD

**Issue by liruilong940607** [kwea123/ngp_pl#22]
Thanks for this great work!

After going through the code, I noticed this two lines behave weirdly:

https://github.com/kwea123/ngp_pl/blob/091c2bc3cd972fe4ccb874749cdd7555590d635a/models/csrc/raymarching.cu#L215-L216

The original Instant-NGP code is [here](https://github.com/NVlabs/instant-ngp/blob/caf9c27b04f9917af5fd8e9664c9ad070e5155b0/src/testbed_nerf.cu#L312):

```c
float mip_scale = scalbnf(1.0f, -mip);
```

Which should be equivalent to 

```c
const float mip_bound = 1 << mip; 
const float mip_bound_inv = 1.0f/mip_bound; 
```

I'm not sure what's the reason of adding a `fminf` operation here.

**Comment by kwea123** [comment]
Hi, this part is still under debugging, so I can't tell you if this code is right or wrong for now, but the idea is as follows: `1<<mip` is an upper bound of `scale`
In instant-ngp, the scale (`aabb_scale` in their code) is always a power of 2, but in my code (and also torch-ngp, since I borrowed his idea) we don't have this assumption. For example, we can have `scale=1.5` which is not a power of two. In this case, the largest `mip` is `1`, but since the scale is only 1.5 not 2, if we fill the 128^3 density grid in 2^3, many cells are wasted. Therefore, we clip the upper limit to `scale` to maximize efficiency.

If this is still unclear or strange to you, I think you can force the `scale` to be a power of 2, then you can apply original instant-ngp code

**Comment by liruilong940607** [comment]
Thanks for your clarification! Yeah that makes a lot of sense. In the case of nerf synthetic scene, since by default your density grid at `mip=0` covers [-1, 1] centered at 0, the model `scale=0.5` means the to set density grid at `mip=0` to cover [-0.5, 0.5], which equivalent to [0, 1] in the ngp's impl. Am I understand correctly?

**Comment by kwea123** [comment]
Actually I fixed the code just now, `mip=0` covers [-0.5, 0.5] (not [-1, 1]), which is equivalent to [0, 1] like you said

https://github.com/kwea123/ngp_pl/blob/585cdba53abf794e3c23b0716c84f22c969f0c7c/models/csrc/raymarching.cu#L165
